crates/agglayer-contracts/src/lib.rs

```diff
@@ -145,35 +157,146 @@ pub enum L1RpcError {
     CacheLockPoisoned,
 }
 
+/// Finds the block number where a contract was deployed using binary search.
+/// This is used to optimize event queries by starting from the deployment
+/// block.
+async fn find_contract_deployment_block_number<RpcProvider>(
+    rpc: &RpcProvider,
+    address: Address,
+) -> Result<Option<u64>, L1RpcError>
+where
+    RpcProvider: alloy::providers::Provider,
+{
+    use alloy::eips::{BlockId, BlockNumberOrTag};
+
+    // Get the latest block number as the upper bound for search
+    let latest_block = rpc.get_block_number().await.map_err(|e| {
+        L1RpcError::FailedToQueryEvents(format!("Failed to get latest block number: {}", e))
+    })?;
+
+    debug!("Finding contract deployment block for address: {address}, latest_block {latest_block}");
+
+    let mut hi = latest_block;
+    let mut lo = 0u64;
+
+    // Quick exit: if no code at latest block, contract was never deployed
+    // (Note: this will treat self-destructed contracts as "not found")
+    let latest_code = rpc
+        .get_code_at(address)
+        .block_id(BlockId::Number(BlockNumberOrTag::Number(hi)))
+        .await
+        .map_err(|e| {
+            L1RpcError::FailedToQueryEvents(format!("Failed to get code at latest block: {}", e))
+        })?;
+
+    if latest_code.is_empty() {
+        debug!("No code found at latest block for address: {}", address);
+        return Ok(None);
+    }
+
+    debug!(
+        "Contract has code at latest block, starting binary search from block 0 to {}",
+        hi
+    );
+
+    // Binary search for the first block where code is present
+    let mut answer: Option<u64> = None;
+    let mut iterations = 0;
+    const MAX_ITERATIONS: u32 = 256; // Safety limit to prevent infinite loops
```

Comment:
We should be able to get the exact block within 25 iterations. Even if limited to just 22, it should get us within some 6 or so blocks of the real answer which is probably good enough.